platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java

```diff
@@ -198,62 +141,189 @@ public Hash hash(
+            }
+            super.completeExceptionally(ex);
+        }
+
+        @Override
+        protected boolean exec() {
+            try {
+                assert count.get() == 0;
+                final Hash hash;
+                if (leaf != null) {
+                    hash = CRYPTO.digestSync(leaf);
+                    listener.onLeafHashed(leaf);
+                    listener.onNodeHashed(path, hash);
+                } else {
+                    int len = 1 << height;
+                    long rankPath = path;
+                    for (int i = 0; i < height; i++) {
+                        rankPath = Path.getLeftChildPath(rankPath);
+                    }
+                    while (len > 1) {
+                        for (int i = 0; i < len / 2; i++) {
+                            final long hashedPath = Path.getParentPath(rankPath + i * 2);
+                            Hash left = ins[i * 2];
+                            Hash right = ins[i * 2 + 1];
+                            if ((left == null) && (right == null)) {
+                                ins[i] = null;
+                            } else {
+                                if (left == null) {
+                                    left = hashReader.apply(rankPath + i * 2);
+                                }
+                                if (right == null) {
+                                    right = hashReader.apply(rankPath + i * 2 + 1);
+                                }
+                                ins[i] = hash(hashedPath, left, right);
+                                listener.onNodeHashed(hashedPath, ins[i]);
+                            }
+                        }
+                        rankPath = Path.getParentPath(rankPath);
+                        len = len >> 1;
+                    }
+                    hash = ins[0];
+                }
+                out.setIn(getIndexInOut(), hash);
+                return true;
+            } catch (final Throwable e) {
+                completeExceptionally(e);
+                throw e;
+            }
+        }
+
+        static Hash hash(final long path, final Hash left, final Hash right) {
+            final long classId = path == ROOT_PATH ? VirtualRootNode.CLASS_ID : VirtualInternalNode.CLASS_ID;
+            final int serId = path == ROOT_PATH
+                    ? VirtualRootNode.ClassVersion.CURRENT_VERSION
+                    : VirtualInternalNode.SERIALIZATION_VERSION;
+            final HashBuilder builder = HASH_BUILDER_THREAD_LOCAL.get();
+            builder.reset();
+            builder.update(classId);
+            builder.update(serId);
+            builder.update(left);
+            builder.update(right);
+            return builder.build();
+        }
+
+        private int getIndexInOut() {
+            if (path == 0) {
+                return 0;
+            }
+            long outPath = out.path;
+            for (int i = 0; i < out.height; i++) {
```

Comment:
This loop could and should be avoided. `getLeftChildPath(path)` is just `(path << 1) + 1` or equivalently `(path << 2) - 1`. Child's path at depth n is thus `(path << (n + 1)) - 1`. This can be added to `Path` as a new API: `Path.getLeftChildPath(path, depth)`.

Reply:
I introduced two new methods to `Path`: to get grand-parent path and to get left-most grand-child path, both are used in virtual hasher. Unit tests for both new methods are added, too.